### ISSUE TITLE
Carrion spawning from corpses

### ISSUE THREAD

**Issue by acidia** [CleverRaven/Cataclysm-DDA#24021]
Now that rotting food has started to generate vermin, what should we do with rotting corpses from a canon point of view?

Different types:
Rotting clean corpses (a deer for example)
Rotting pulped clean corpses (a deer you ran over)
Rotting pulped tainted corpses
Rotting NPC/player corpses

Do we want all clean corpses to spawn carrion monsters (assuming the corpse is eaten)?
Should a pulped zombie generate an amalgamation creature (like a little jabberwock), a blob type creature, or should they get eaten also?  Or should a pulped creature eventually come back together to become the original creature again?

**I-am-Erk edit 2019**: I'm bumping this topic because it has a lot of good discussion, and adding a 2019 update here: https://github.com/CleverRaven/Cataclysm-DDA/issues/24021#issuecomment-466097887

**Comment by DemAvalon** [comment]
If my opinion is worth anything, I would say that pulped corpses should not turn into anything, the whole point of pulping a corpse is to stop it from rising so if pulping a crop generates more monsters or reforms itself, people will simply not pulp anymore and only butcher and burn the meat (since corpses take a while to burn and the meat is easy to burn), which seems kind of tedious to have to not only butcher the zombies but also to burn the remains so they don't spawn more annoyances, so this change kinda doubled the time one has to spend pulping and butchering corpses.

**Comment by NotFuji** [comment]
Lets also not forget that zombie corpses will exist in the hundreds usually.  Spawning even one carrion per corpse is going to be a significant performance hit.

Furthermore if pulped zombies just turn into a more dangerous creature eventually anyway, there's not much point to pulping at all in that case.  And again, the possibly hundreds of zombie corpses lying around means cleaning them up to the point at which they won't be a problem takes an annoying amount of time.  Not even considering the time required to round up all the newly butchered meat for burning so exponentially larger numbers of carrion don't get spawned anyway.

Non-tainted corpses are few enough that there isn't much of a problem there, though.

**Comment by acidia** [comment]
Doesn't it seem crazy though that nothing would come from a giant pile of dead bodies just because you hacked them up?  For Romero zombies, I could understand... but even then, giant flocks of vultures or wild dogs would thrive on a massive pile of meat.  A tide of football sized maggots should flood the area within a week.  It just feels like areas are FAR too easy to control with a single person, then no new threats or problems will ever exist.

**Comment by DracoGriffin** [comment]
>Doesn't it seem crazy though that nothing would come from a giant pile of dead bodies just because you hacked them up? For Romero zombies, I could understand... but even then, giant flocks of vultures or wild dogs would thrive on a massive pile of meat. A tide of football sized maggots should flood the area within a week. It just feels like areas are FAR too easy to control with a single person, then no new threats or problems will ever exist.

Well there is wandering horde settings already for that.

Secondly, one of the main issues people have with fungaloids (and deriatives, myself included) is they become a HUGE issue on performance, where movements and actions start taking real life seconds or longer to perform (like moving a tile will be delayed by 5-15 seconds in real life before it completes). This could very easily compound that issue further and make all the work moot (although there is the option/"mod" to remove fungal enemies).

I really enjoy the aspect you're reaching for (as well as the options that can be finagled out of it [like a skeletal version of animal husbandry/breeding]) but I think it's much better to do small steps, rather than trying to rapidly backtrack/revert. And considering CDDA is meant to be single-player (and NPCs aren't really viable as alternatives yet), it's a bit difficult to expect players to have to exert that much influence and micromanagement (although for the multiplayer Looming Darkness that may be more feasible).

My opinion is that pulped should be limited to very small amounts of vermin/wildlife (treat it similarly to wandering hordes -- except instead of a point system, a "rotten item" system so long as there are X number of rotten items/types, there will also generate Y number of vermin/zombies/wildlife/nether creatures/etc [with a hard cap, for example 10 pulped corpses makes 20 or so smaller vermin types or 2-5 larger nether creature types, like some sort of cost multiplier] until there is no rotten items to generate from). So you would still get your goal (leaving waste and dead bodies inevitably brings along baddies) without compromising the player's computer with exponential growth ( dead stuff = creates monsters, created monsters kill other created monsters or stuff = more dead stuff, ad infinitum).

**Comment by voxinaudita** [comment]
I see a lot of games suffer when the devs decide that something is "too easy". Reactive design like this can result in gameplay which is complicated rather than complex, if that makes sense. Will the changes result in player action which is simply repetitive (like pulping corpses is now, it makes "sense" but is often just busywork), or result in the player coming up with novel ways to solve a new problem? I think it's the former.

**Comment by DemAvalon** [comment]
> Doesn't it seem crazy though that nothing would come from a giant pile of dead bodies just because you hacked them up? For Romero zombies, I could understand... but even then, giant flocks of vultures or wild dogs would thrive on a massive pile of meat. A tide of football sized maggots should flood the area within a week. It just feels like areas are FAR too easy to control with a single person, then no new threats or problems will ever exist.

If a player can single handedly kill a hundred zombies, they can kill anything that would be inclined to approach the carnage, rotting zombie corpses would not generate monsters or a credible threat to the player, it would generate busywork

**Comment by acidia** [comment]
If a player can single handedly kill a hundred zombies without issue, there won't be a problem with carrion spawns because the player has already beaten the game and likely won't be playing much longer.  

For everyone else though, I would like to see the monster ecosystems rebalance to a post-zombie world.  The number of zombies in the world should pretty much only be decreasing given that just about all of the people are dead.  With the carrion growth PR, we will have closed ecosystems (pending balancing).  Creatures that eat zombies (nether?) should probably start becoming common/dominant year 2+.  If we as a whole don't want ecosystems to form/change we can scrap everything right now though.

**Comment by NotFuji** [comment]
> If a player can single handedly kill a hundred zombies without issue, there won't be a problem with carrion spawns because the player has already beaten the game and likely won't be playing much longer.

That's exactly the problem though, as others have stated.  In that case the carrion serve as nothing but a no-fun annoyance.

Also, being able to kill 100 zombies far from "beaten the game."  A new-game player out of the shelter with zero skills and a pistol can do that.  Melee with a long stick and some added tactics.  Trivial with a working car.  Zombies are the baseline early-game enemy, far before harder stuff like triffids and fungus.

Changing ecosystems is perfectly fine.  But what system controls that and how it's implemented needs to take enjoyability into account first and foremost, not "it's too easy."

**Comment by voxinaudita** [comment]
Game "realism" is pretty subjective, and people can argue back and forth for ages about it. Instead, can you pose new design ideas in terms of goals for player interaction? Is the purpose of these ecosystems to give players new challenges and solve new puzzles? Will it give players more choices, or will it take choices away?
Also, how well will this fit with the rest of the game? Another problem a lot of games get is that they will throw in a new enemy type (or new puzzle, terrain, obstacle, etc.) which can only be defeated with another new item which is introduced at the same time. A few of these changes later and the player now has to carry a huge amount of different special-use items everywhere.

**Comment by DemAvalon** [comment]
> If a player can single handedly kill a hundred zombies without issue, there won't be a problem with carrion spawns because the player has already beaten the game and likely won't be playing much longer.

I mean no disrespect, but you seem to have a very small threshold of what the end game is, a regional school has hundreds of zombies, and is one of the first places an experienced survivor will raid (for books)

a changing ecosystems would be a neat idea, but the idea being discussed right now is whether zombie corpses (and tainted meat) should cause the spawning of carrion, which will cause wasted processing on harmless creatures, and whether doubling the time one has to dedicate to clear an area is worth spawning said creatures.

which can all easily be handwaved by saying that animals are repulsed by tainted meat because the smell or taste is revolting to them, after all it is poisonous, so only food which is safe for consumption raw meat included would generate carrion.

**Comment by acidia** [comment]
> I mean no disrespect, but you seem to have a very small threshold of what the end game is, a regional school has hundreds of zombies, and is one of the first places an experienced survivor will raid (for books)
> 

Tell me more, please.  When I was building the mall, to a far lesser extent when I built the regional school, I had to listen to SO MANY people complain that I was completely ruining their game by creating fixed structures instead of having interior rooms be randomly generated.  I just wanted to tell people to STFU and try the thing first but the complaining didn't stop till about a month after the buildings were added.  I think that may predate even our previous forum...

> idea being discussed right now is whether zombie corpses (and tainted meat) should cause the spawning of carrion, which will cause wasted processing on harmless creatures, and whether doubling the time one has to dedicate to clear an area is worth spawning said creatures.
>
Wait, why are carrion spawns harmless?  They eat your food, your crops, their biosignatures attract larger predators to your location, and if left unchecked they can make areas that were once safe dangerous again by growing and breeding.  This all takes place over two or three seasons to get into full swing so I'm a little confused as to how its already become such a burden on your processing power unless you cranked carrion spawns way up (which you can also just turn off as a global option) or what build you are testing it.

Please let me know what version you are using, how many days you are in, and your observations.
Don't forget to get all of the different PRs! #23833, #23930, #24020

**Comment by Zireael07** [comment]
Rivet is still around, I think she was around when you first showed up, too.

**Comment by acidia** [comment]
Ya!  I think we started on whale's forum at about the same time but I'm pretty sure I got my first PR up first :D She pretty much has the longest PR streak of anyone alive though.

**Comment by BorkBorkGoesTheCode** [comment]
Rivet left an emoji yesterday. She seems to prefer merging nightly PR content right now.

**Comment by DracoGriffin** [comment]
@acidia Totally an aside, but you should check out the unofficial Discord -- there is a development channel where some more discussion takes place and could be useful for feedback.

**Comment by NickGrease** [comment]
First, I got to say this is a really great idea/concept. There seem to be a lot of possibilities for new game play. As for what should spawn carrion, I'd say everything has a chance; blood, bone, and flesh tainted or not. If you're playing with revivification ON, chances are you're doing it because you want the consequences of not cleaning up. This adds one more. This *could* be considered busy work, but I like to think it leaves room for new mechanics; for an off the top of my head example, instead of butchering or pulping a body, maybe a splash of bleach/ammonia/lye would prevent critters from spawning but still allow the resurrection of un-pulped/non-butchered corpses.

Another option, to control insane spawn rates, might be for carrion that do spawn, to seek out other corpses/meat and eat it, preventing further spawns, but also growing in strength.

Another idea, again just spit-balling, and to further reduce the chance of carrion spawning out of control, a story; I kill 10 zeds outside my place, and being tired/lazy I call it a night, leaving the corpses out there. In the morning I wake to find one corpse has spawned a critter (corpse removed) and 5 more are missing. I go out hunting for breakfast, and instead of 10 turkeys spawning, 5 normal turkeys spawn and 5 mutated ones spawn. The idea being that during the night there's a chance for wildlife to 'remove' corpses and then becoming 'something else' instead of normal wildlife spawns. Maybe this would affect a few map tiles - with the driving force being 'clean up your mess or edible wildlife becomes impossible to find in that area.'

**Comment by ZhilkinSerg** [comment]
> I mean no disrespect, but you seem to have a very small threshold of what the end game is, a regional school has hundreds of zombies, and is one of the first places an experienced survivor will raid (for books)

Lol, it was @acidia who [added that school](https://github.com/CleverRaven/Cataclysm-DDA/pull/864).

Actually you better show some respect for [their hard work](https://github.com/CleverRaven/Cataclysm-DDA/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Aacidia+is%3Aclosed+is%3Amerged).

**Comment by DemAvalon** [comment]
@acidia, you were right, I did some testing (At least I think I did) like you said with all the pulls (even though my main concern was only the plague nymphs and other possible tainted carrion)

I went to a regional school and killed about 30 zombies and butchered them, I remained in the area (right beside the rotten tainted meat) for a few days and nothing spawned, leaving the area (teleporting far, then returning) showed that only 3-4 plague nymphs spawned, all the tainted meat was gone.

Next I piled stacks of 1000 tainted meats (in a different area far from there) for a total of 10000 tainted meats (10 stacks in ten squares), left for a day, and upon returning there were only 3-4 plague nymphs there.

Next I spawned many zombies and splattered them (with Mjonir) so there were several piles of tainted meat spread across the area (many stack of 1-2 meat), about twenty-eight stacks, then left again and waited, upon returning there were many plague nymphs (about 15-20) so it seems the checks to spawn carrion happen for individual stacks not total units (which was my biggest fear) so in theory a single stack of 1000 zombie corpses would still generate a very small amount of plague nymphs.

Plague nymphs are annoying in the sense that they are quick and flee from you, plus zombies ignore them for some reason, also they do not eat any other type of food only tainted meat, tested this by having a variety of food around them (raw clean meat, candies, spaghetti, etc.) and only the tainted meat was consumed.

In conclusion, their numbers are not as vast as I thought, and it seems like it takes a long time for them to become a problem, I maintain, however, that pulped zombies should remain pulped, if another creature comes and eats it that it's fine as long as it does not rise again and that the plague nymphs are harmless and a bit annoying, and that this might increase the time one spends clearing an area, but the results of simply not doing the extra work are not aggravating as I thought.

So @acidia, I apologize and promise I will learn from my error, I swear that I will endeavor to test things properly before giving feedback, once again I apologize deeply and sincerely for my ignorant assumptions.

and @ZhilkinSerg my point about that was that a player facing and killing a hundred zombies is not an endgame situation, and I was sincere in the part I said I meant no disrespect, I genuinely meant no harm or offence by my words, but it seems that my poor wording may have caused the effect none the less, so I apologize for my poor wording, and I assure you that I had no intent to  cause offense or to mock @acidia in any way, shape, or form.

**Comment by acidia** [comment]
> Next I piled stacks of 1000 tainted meats (in a different area far from there) for a total of 10000 tainted meats (10 stacks in ten squares), left for a day, and upon returning there were only 3-4 plague nymphs there.
> 
Yup, this is something I'm aware of.  I looked at changing it but it tended to cause more problems than anything because it kept trying to spawn creatures through walls (excessively) if I gave it more than 0 range.  Items that come in giant stacks that are small portion sizes also tended to gen monsters far too heavily.  Something that should get addressed but is more of a balance thing if it is needed.

> Plague nymphs are annoying in the sense that they are quick and flee from you, plus zombies ignore them for some reason, also they do not eat any other type of food only tainted meat, tested this by having a variety of food around them (raw clean meat, candies, spaghetti, etc.) and only the tainted meat was consumed.
> 
Zombies should ignore them since they are considered tiny wildlife until they grow up to the second stage.  If only meat is disappearing then it is due to rot, not consumption.  I couldn't add the special attack flag to the JSON entry without the two PRs being combined or conflicting.  I did most my testing with flour and tomatoes (just because charges weren't working correctly in the first run).  They'll eat the food right in front of you while you watch if everything is correct.  I'll double check, might need a game message that says they ate something.

> So @acidia, I apologize and promise I will learn from my error, I swear that I will endeavor to test things properly before giving feedback, once again I apologize deeply and sincerely for my ignorant assumptions.
> 
100% cool man I wasn't as offended as it sounded, I just thought it was funny that you chose one of my own buildings that I got shit for making years ago.  Thanks a bunch for testing also!

**Comment by DemAvalon** [comment]
> Zombies should ignore them since they are considered tiny wildlife until they grow up to the second stage. If only meat is disappearing then it is due to rot, not consumption. I couldn't add the special attack flag to the JSON entry without the two PRs being combined or conflicting. I did most my testing with flour and tomatoes (just because charges weren't working correctly in the first run). They'll eat the food right in front of you while you watch if everything is correct. I'll double check, might need a game message that says they ate something.

Huh, I didn't know they ignored giant cockroach nymphs too, seems weird to me that they ignore those because they are tiny wildlife (the description mentions that they are the size of a rat) but will attack black rats and sewer rats (maybe the description should emphasize that they are large rats, but then again maybe that is common knowledge and I'm being stupid as usual)

Also, it seems they ignore the large versions as well (skittering plagues and plague vector), as for them eating normal food, will their need to eat supersede their fear of the player? (like if the food is right next to you will they approach to eat?), and could a dog tied in place keep them away and act well... as a guard dog?, if they are more interested in eating than running away, I imagine one could use food as a lure to kill them, or lure them into traps with food.

And thank you for your patience with me.

**Comment by John-Candlebury** [comment]
I think its a great mechanic. 

However theres a problem no one has mentioned, this mostly afects people who backtrack their steps, but I feel that very few people if at all do this. This could be fixed with some long-term features like better npcs and fast travel, but I think something should be done in the meantime.

So I think there should be some work done in making returning to cleared areas somehow desirable. Maybe make some of the new critters worth hunting, say let the stronger roaches dop glands full with mutagen or some other useful chemicals.

>Creatures that eat zombies (nether?) should probably start becoming common/dominant year 2+. If we as a whole don't want ecosystems to form/change we can scrap everything right now though.

I think we are better of with the lovecraftian stuff being  much more powerfully and very very rare. Limited to  special map gen locations and requiring the player to specifically prepared to handle them. 

They make cool enemies but having them be the standard enemies would be cheap, they should never be easy type of foe to defeat.

**Comment by thethunderhawk** [comment]
>they should never be an easy type of foe to defeat

Mi-Gos and Krecks are pretty easy to defeat. I always envisioned this as an ongoing cross-dimensional invasion (as per that super old design doc). It makes sense to me that over time they should become more common and the line between zed/blob/nether creature would become ever harder to define as the years wore on. In that sense I think that they should indeed be super rare and relegated only to special areas, but I could see carrion spawn-related nether creatures in the distant lategame.

**Comment by kevingranade** [comment]
I'm pretty torn on this one, I really like the idea of zombie meat left lying around causing trouble, but I don't like subverting the current "kill and butcher is sufficient" status quo.
I really don't like pulped bodies reassembling themselves.
I don't like it spontaneously generating new aggregate creatures, with a possible exception of a "critical mass" situation where the player has piled up an extremely large mass of zombie corpses.

Basically I think they should get eaten the same as other corpses, though perhaps generating different sets of scavengers.

I'm not worried about performance of this, but to avoid just being annoying by spawning huge numbers of weak enemies, it should progress rapidly to generating a small number of dangerous creatures.

**Comment by nexusmrsep** [comment]
Side note: I personaly would like to see actual wild life eating leftovers. For normal meat, leftover guts and pieces, some crows, vultures, wolves, coyotes, if not intrested in either eating nearby player or running from a player at the given moment, would not forsake an opportunity of lunch in the harsh time of Cataclysm where other food is more scarce than it once was.

For zombie meat and leftovers i was thinking of it as maybe becoming food for other zombies when they are not otherwise attracted by the player. Although this is a far-shot guess, as i dont know what canon says about blob-eat-blob matters, its obvious that more blob is better for individual blob, so mayhaps they might eat tainted meat to improve themselves. That would also explain in a way how hulks got so big, without breaking laws of physics. 

If that idea is trash, then maybe there are also other, not so picky carrion eaters around, although none come to my mind at the moment.

**Comment by thethunderhawk** [comment]
Here’s a thought (long as we’re talking about ideas that could be trash):
If the blob is part of a extradimentional ecosystem, maybe something feeds on it. Could be fungus, triffid (already partially implemented?), or nether horror. Wouldn’t show up until the later stages of the apocalypse, and perhaps it could even be an indicator that the world is fundamentally changing (into a lategame with advanced zed evos, expanded fungus and triffid presence, new nethers, etc...)

**Comment by thethunderhawk** [comment]
>blob doesn’t seem to be any more toxic than what’s normally generated during putrefaction 

Aside from all that mutagen, right? IMO things that eat blob might start out as IRL carrion eaters, but even with the best possible digestive system it’s like chugging Tainted Tornado breakfast lunch and dinner. Things really ought to get weird.

**Comment by FulcrumA** [comment]
@thethunderhawk 
I don't recall meat itself to carry much of mutagenic properties. Maybe I didn't eat enough of it. As far as I know it's once processed that it can mess with one's genes. And even if it'd be mutagenic, it's not certain it works on said carrion eaters, isn't digested harmlessly etc as many such creatures already can eat plenty of things deadly to people and survive unscathed - so it's really a matter of deciding how much worse such mutagen is and how it would affect wild fauna in regards to which I mentioned before - it doesn't seem more toxic than regular rotting meat. It'll poison you, yeah, but won't immediately kill you.

**Comment by thethunderhawk** [comment]
Tainted meat contains enough mutagenic properties that simply blending it with whiskey results in a low level mutagen. You can extract the blob from tainted meat, and you can eat mutated arms and legs for a direct mutagen effect. I agree that carrion eaters *could* be immune to that, but I think there could be room for some flavor in there too, especially if those dangers somehow could be exclusive to the mid-late game.

**Comment by I-am-Erk** [comment]
I'm bumping this topic and adding it to farming improvements.
Currently the state of vermin is complex. There are a few issues to look at. In my opinion:

1. Vermin spawns should reduce the source material by the mass of an adult vermin, to prevent infinite spawns. Small fractions could still produce a vermin to make sure they come out in numbers (ideally maybe looking around for other sources of vermin and taking from those sources to make up the difference.
2. "Vermin" spawns that are not bugs should be removed, to prevent wolves spawning inside your codoned-off farm.
3. All vermin spawns should be carefully considered for gameplay value. I recommend making them much, much less common in addition to reducing the infinite generation problem, so that giant vermin only really appear in huge sources of rotting flesh eg. cities
4. Vermin should mostly be moved to places like the sewers.

**Comment by voxinaudita** [comment]
> All vermin spawns should be carefully considered for gameplay value.

Yes, please. I'm not sure what the original motivation was for vermin. In terms of realism, vermin are treated as life arising from the old theory of spontaneous generation, e.g. it was once believed that a "recipe" for mice was scattering grain and covering it with straw. "Vermin" in real life play a large role in a complex ecosystem.

In terms of gameplay, vermin are still busywork. What gameplay problem was solved by adding them? Was it something like "players aren't putting enough thought into storing and protecting their food"?

Lastly, in an apocalypse, the sewers would be absolutely devoid of food sources for the fauna that normally occupy this habitat. Storm drains would still route water, but nobody is around to use the toilets and sinks where sewage comes from. Anything that consumes decaying material now has a feast waiting for them on the surface.

Actually lastly, cockroaches aren't even active during cold weather.

**Comment by l29ah** [comment]
Oh shit i killed a handful of zombies and now i have a cockroach farm for free!

**Comment by I-am-Erk** [comment]
For the moment, vermin spawns have been disabled completely. I'm going to close this issue for now because I think the problem is going to be totally rethought.
